Add UpdateInstantiateConfig gov proposal

jhernandezb opened the pull request:
closes #780 
closes #810




- [x] Server Side
- [x] Client Side
- [x] Testing

jhernandezb commented:
@alpe  Regarding single update vs multiple updates on a single proposal. Should it be one proposal  at a time? or udpating multiple is fine? also different configs? Currently I have multiple configs with the same AccessConfig.

Also I was unable to run `make proto-all` only `make proto-gen` worked the other commands are giving me errors of proto definitions not found. (`lint, proto-breaking`)

alpe commented:
@jhernandezb good start! You are more familiar with the concrete use case but from my perspective multiple code_ids with same AccessConfig make a lot of sense. This can be explained to the voters but also gives flexibility to address multiple code ids.

jhernandezb commented:
Yes my reasoning is for example: Freeze Code ID 1,2,3,4 then you only need to apply the same config to all of them. If someone wanted different access configs then a different proposal should be made explaining the reasoning.

Will complete this PR then and push updates

ethanfrey commented:
This will close #780 right?

If so, please add that to the description

ethanfrey commented:
> @alpe Regarding single update vs multiple updates on a single proposal. Should it be one proposal at a time? or udpating multiple is fine? also different configs? Currently I have multiple code ids with the same AccessConfig.

I would either:
* One code Id, one instantiate config
* A list of N repeated (code_id, instantiate_config) submessages

I have no strong opinion either way. Although I guess the second would be better to match the use case you had in mind to allow multiple code_ids

ethanfrey commented:
If we can get this ready by (early?) next week, I would love to get it in the upcoming v0.27 release

jhernandezb commented:
working on this today can have a PR ready for review tomorrow so I think its doable to have it for v0.27

jhernandezb commented:
@ethanfrey @alpe this is ready for review will be doing some manual testing.

Some notes:
- Added a new method to the keeper `SetAccessConfig`
- added testing to proposal_integration tests with different cases but let me know if it can be improved
- Moved to `A list of N repeated (code_id, instantiate_config) submessages` approach 
- CLI command is a bit tricky with that approach but added examples

alpe reviewed (changes requested):
Excellent work. I am very happy with this PR.

Only some comments on rest that need be addressed before merging.

Too bad we have not dropped legacy rest support already.

jhernandezb commented on x/wasm/client/rest/gov.go:
```diff
@@ -407,7 +407,50 @@ func (s UnpinCodeJSONReq) GetBaseReq() rest.BaseReq {
+
+	CodeUpdates []types.CodeAccessConfigUpdate `json:"code_updates" yaml:"code_updates"`
+	// InstantiatePermission to apply on contract creation, optional
+	InstantiatePermission *types.AccessConfig `json:"instantiate_permission" yaml:"instantiate_permission"`
```
good catch it doesn't need to be here.

jhernandezb commented on x/wasm/client/rest/gov.go:
```diff
@@ -407,7 +407,50 @@ func (s UnpinCodeJSONReq) GetBaseReq() rest.BaseReq {
+	Proposer    string    `json:"proposer" yaml:"proposer"`
+	Deposit     sdk.Coins `json:"deposit" yaml:"deposit"`
+
+	CodeUpdates []types.CodeAccessConfigUpdate `json:"code_updates" yaml:"code_updates"`
```
this follows the protobuf defitinion 
```
repeated CodeAccessConfigUpdate code_updates = 3
```

should we rename the proto field to code_access_config_updates ? 

I think it would make sense because is just doing an specific action and not performing a full update to a code id.

thoughts?

ethanfrey replied:
I think `access_config_updates` or `code_access_config_updates` would work (and ideally update both in the Protobuf file as well)

jhernandezb commented on x/wasm/types/exported_keepers.go:
```diff
@@ -53,6 +53,9 @@ type ContractOpsKeeper interface {
 	SetContractInfoExtension(ctx sdk.Context, contract sdk.AccAddress, extra ContractInfoExtension) error
+
+	// SetAccessConfig updates the access config of a code id.
+	SetAccessConfig(ctx sdk.Context, codeID uint64, config AccessConfig) error
```
`NewWasmProposalHandlerX` takes `ContractOpsKeeper` as argument, not really sure how have access to this method without exporting it (not really sure if a private method makes sense on an interface)

ethanfrey replied:
good point here. If there is a simple way to keep it private and expose to the proposal handler, maybe @alpe can make a quick PR to show how. Otherwise, I would be fine to merge like this.

jhernandezb commented on x/wasm/keeper/proposal_integration_test.go:
```diff
@@ -749,3 +749,101 @@ func TestUnpinCodesProposal(t *testing.T) {
+		everybody   = StoreRandomContractWithAccessConfig(t, ctx, keepers, &mock, &types.AllowEverybody)
+		withAddress = StoreRandomContractWithAccessConfig(t, ctx, keepers, &mock, &withAddressAccessConfig)
+	)
+	type codeUpdate struct {
```
refactored to reuse that type.

jhernandezb commented on x/wasm/client/rest/gov.go:
```diff
@@ -407,7 +407,50 @@ func (s UnpinCodeJSONReq) GetBaseReq() rest.BaseReq {
+}
+
+func UpdateInstantiateConfigProposalHandler(cliCtx client.Context) govrest.ProposalRESTHandler {
+	return govrest.ProposalRESTHandler{
```
fixed

jhernandezb commented:
Made some updates just have couple questions
https://github.com/CosmWasm/wasmd/pull/796#discussion_r866928010
https://github.com/CosmWasm/wasmd/pull/796#discussion_r866932343

ethanfrey commented:
@alpe does this address all your concerns?

I will do a review and merge this evening / tomorrow unless there is a blocker

ethanfrey commented on proto/cosmwasm/wasm/v1/proposal.proto:
```diff
@@ -148,3 +148,25 @@ message UnpinCodesProposal {
+  // CodeID is the reference to the stored WASM code to be updated
+  uint64 code_id = 1 [ (gogoproto.customname) = "CodeID" ];
+  // InstantiatePermission to apply to the set of code ids
+  AccessConfig instantiate_permission = 4 [ (gogoproto.nullable) = false ];
```
Odd that this is 4. Not wrong, just surprising, I assume this was for some backwards compatibility with another struct?

It would be good to label it if so, with something like "2 and 3 are reserved" or such

jhernandezb replied:
it should be 2 will fix.

ethanfrey commented on x/wasm/client/cli/gov_tx.go:
```diff
@@ -614,3 +616,109 @@ func ProposalUnpinCodesCmd() *cobra.Command {
+	cmd.Flags().String(cli.FlagDeposit, "", "Deposit of proposal")
+	cmd.Flags().String(cli.FlagProposal, "", "Proposal file path (if this path is given, other proposal flags are ignored)")
+	// type values must match the "ProposalHandler" "routes" in cli
+	cmd.Flags().String(flagProposalType, "", "Permission of proposal, types: store-code/instantiate/migrate/update-admin/clear-admin/text/parameter_change/software_upgrade")
```
I'm a bit confused why this flag is here?

Usage is `tx gov submit-proposal update-instantiate-config 1,nobody 2,everybody 3,%s1l2rsakp388kuv9k8qzq6lrm9taddae7fpx59wm` right?

ethanfrey replied:
Good point, can you make a follow up (cleanup issue), so someone can look into that later.
Not blocking merge at all, but don't want to forget it

ethanfrey commented on x/wasm/keeper/proposal_integration_test.go:
```diff
@@ -749,3 +749,98 @@ func TestUnpinCodesProposal(t *testing.T) {
 	}
 }
+
+func TestUpdateInstantiateConfigProposal(t *testing.T) {
```
nice

ethanfrey reviewed (commented):
lgtm. I would love to see a manual test (as there are all kinds of codecs and middleware called in the real proposal stack), but good stuff.

I added a few comments on naming, which would be nice polish.
A CHANGELOG entry (one line pointing to this PR) would also be nice.

Happy to approve once those are done (just mention me in a comment and I will double check)

jhernandezb commented:
@ethanfrey addressed your comments and also here is a demo
https://www.loom.com/share/e99721a095d4428d80f81496daaee2ef
(ignore the print long on submit proposal, I had that as debug on my local cosmos-sdk dependency)

ethanfrey reviewed (approved):
Great stuff and loving the loom video!

Thank you very much
